### DISCUSSION TITLE
Shortcode to insert all the images from the subfolder

### DISCUSSION THREAD

**Discussion by jurekure** [11ty/eleventy#2192] [Q&A] [answered]
I hope the noob questions are allowed also.
I'm trying to get this shortcode working, which would go through specific folder and output all the pictures files in that folder. im really not fluent in JS but i think that from the code below the output that this code is producing is kinda on the right track, but 11ty still produce "undefined" instead of rendered shortcode. I had tried with some easier example with only return function and that works without a hiccup. Again maybe the only problem is my lacking knowledge of JS, but still hope you can set me on the better path.

`
  config.addShortcode("photoFolder", function (photoFolder) {

    const photoList = []

    fs.readdir(photoFolder, (err, files) => {
      
      if (err) throw err;
      if (files) {}
      files.forEach(file => {

        photoList.push(`<img src=${photoFolder}/${file}ffdfdfdffddfd>`);

      });
      return photoList.join("/n");
    });
  });`

**Comment by pdehaan** [comment, marked as answer]
I think you're close!

The shortcode needs to return the array of files (you're returning early in the `fs.readdir()` call).
But even if we move the `return photoList.join("\n")` down slightly, I think you'll still have an issue w/ the async nature of `fs.readdir()`. See the [bad] code below:

```js
// DO NOT COPY/USE THIS CODE.
config.addShortcode("photoFolder", function (photoFolder) {
  const photoList = [];
  fs.readdir(photoFolder, (err, files) => {
    if (err) throw err;
    files.forEach((file) => {
      photoList.push(`<img src="${photoFolder}/${file}" />`);
    });
  });
  // THIS WONT WORK because it's returning the empty array before the `fs.readdir()` has finished async reading the directory.
  return photoList.join("\n");
});
```

So our shortcode will create an empty photolist, then `readdir()` will async read the files and populate the `photoList[]` array. But, before that happens, the method has already returned the empty array and our template will have no output.

Here's my modified code which reads the directory synchronously to avoid the timing issue:

```js
config.addShortcode("photoFolder", function (photoFolder) {
  const html = fs.readdirSync(photoFolder)
    .map((file) => `<img src="${file}" />`);
  return html.join("\n");
});
```

That works for me locally; but could use some extra error handling probably (like the folder doesn't exist, or there are non-image files in that directory, etc).

In fact, here's my input directory structure:

```sh
tree -a src
src/
├── assets/
│   └── images/
│       ├── .DS_Store
│       ├── five.jpeg
│       ├── four.jpg
│       ├── one.png
│       ├── three.gif
│       └── two.png
└── index.njk
```

And my generated output looks like this:

```html
<img src=".DS_Store" />
<img src="five.jpeg" />
<img src="four.jpg" />
<img src="one.png" />
<img src="three.gif" />
<img src="two.png" />
```

So you could add some extra logic around only showing certain file extensions, or adding some sorting (it seems to be sorted by filename, which is why images one-five aren't in an expected order; but that may or may not matter to you depending on your naming scheme).

**Reply by pdehaan** [reply]
Posted my sample repo at https://github.com/pdehaan/11ty-2192 if you wanted to see the whole example.

**Reply by pdehaan** [reply]
And some potential tweaks (filter by files only, and have an allow-list of extensions) in https://github.com/pdehaan/11ty-2192/pull/1 pull request since I had the files open.

**Reply by jurekure** [reply]
This is epic! Thanks for such a extensive reply. Didn't even think that im using async process. Again thanks for extensive explanation. Will go through all so i can understand it step by step 👍
